hearts/api/tests/test_socket_events.py

```diff
@@ -133,3 +133,96 @@ def test_create_game_when_last_player_joins(socket_clients, db):
+    assert len(game['data']['rounds'][-1]['pass_selections']) == 1
+    assert game['data']['rounds'][-1]['pass_selections']['user4'] == cards
+
+
+def test_pass_cards_add_to_pass_selections_2_users(db, socket_clients):
+    room, room_id = create_room()
+
+    usernames = ['user1', 'user2', 'user3', 'user4']
+    clients = [socket_clients.new_client() for _ in range(4)]
+    for username, client in zip(usernames, clients):
+        client.emit('join', {'room': room_id, 'username': username})  # Last user to join starts the game
+
+    room = get_room(room_id)
+    game_id = str(room['game_id'])
+    game = get_game(game_id, deserialize=False)
+    current_round = game['data']['rounds'][-1]
+
+    # user1 and user2 choose to pass their first 3 cards
+    user1_cards = current_round['hands']['user1'][:3]
+    user2_cards = current_round['hands']['user2'][:3]
+    clients[0].emit('pass_cards', {'cards': user1_cards})
+    clients[1].emit('pass_cards', {'cards': user2_cards})
+    log = clients[1].get_received()
+    assert log[-1]['name'] == 'pass_submission_status'
+    assert log[-1]['args'][0]['status'] == 'success'
+
+    game = get_game(game_id, deserialize=False)
+    assert len(game['data']['rounds'][-1]['pass_selections']) == 2
+    assert game['data']['rounds'][-1]['pass_selections']['user1'] == user1_cards
+    assert game['data']['rounds'][-1]['pass_selections']['user2'] == user2_cards
+
+
+def test_pass_cards_invalid_pass(db, socket_clients):
+    room, room_id = create_room()
+
+    usernames = ['user1', 'user2', 'user3', 'user4']
+    clients = [socket_clients.new_client() for _ in range(4)]
+    for username, client in zip(usernames, clients):
+        client.emit('join', {'room': room_id, 'username': username})  # Last user to join starts the game
+
+    room = get_room(room_id)
+    game_id = str(room['game_id'])
+    game = get_game(game_id, deserialize=False)
+    current_round = game['data']['rounds'][-1]
+
+    user1_cards = current_round['hands']['user1'][:2]   # Only 2 cards chosen
+    clients[0].emit('pass_cards', {'cards': user1_cards})
+    log = clients[0].get_received()
+    assert log[-1]['name'] == 'pass_submission_status'
+    assert log[-1]['args'][0]['status'] == 'failure'
+    assert 'You cannot pass' in log[-1]['args'][0]['message']
+
+
+def test_pass_cards_all_users(db, socket_clients):
+    room, room_id = create_room()
+
+    usernames = ['user1', 'user2', 'user3', 'user4']
+    clients = [socket_clients.new_client() for _ in range(4)]
+    for username, client in zip(usernames, clients):
+        client.emit('join', {'room': room_id, 'username': username})  # Last user to join starts the game
+
+    room = get_room(room_id)
+    game_id = str(room['game_id'])
+    game = get_game(game_id, deserialize=False)
+    current_round = game['data']['rounds'][-1]
+
+    user_cards = [current_round['hands'][user][:3] for user in usernames]  # 3 cards from each hand
+    for x in range(4):
```

Comment:
Pro tip: you can also do `for client in clients:`

If you think you need the index, you can usually use zip: `for client, card in zip(clients, user_cards):`

not required, just a fun fact